lib/asar/fs-wrapper.ts

```diff
@@ -0,0 +1,779 @@
+import { Buffer } from 'buffer';
+import * as path from 'path';
+import * as util from 'util';
+
+const asar = process._linkedBinding('electron_common_asar');
+const v8Util = process._linkedBinding('electron_common_v8_util');
+
+const Module = require('module');
+
+const Promise: PromiseConstructor = global.Promise as any;
+
+const envNoAsar = process.env.ELECTRON_NO_ASAR &&
+    process.type !== 'browser' &&
+    process.type !== 'renderer';
+const isAsarDisabled = () => process.noAsar || envNoAsar;
+
+const internalBinding = (process as any).internalBinding;
+delete (process as any).internalBinding;
+
+const nextTick = (functionToCall: Function, args: any[] = []) => {
+  process.nextTick(() => functionToCall(...args));
+};
+
+// Cache asar archive objects.
+const cachedArchives = new Map();
+
+const getOrCreateArchive = (archivePath: string) => {
+  const isCached = cachedArchives.has(archivePath);
+  if (isCached) {
+    return cachedArchives.get(archivePath);
+  }
+
+  const newArchive = asar.createArchive(archivePath);
+  if (!newArchive) return null;
+
+  cachedArchives.set(archivePath, newArchive);
+  return newArchive;
+};
+
+// Separate asar package's path from full path.
+const splitPath = (archivePathOrBuffer: string | Buffer) => {
+  // Shortcut for disabled asar.
+  if (isAsarDisabled()) return { isAsar: false };
+
+  // Check for a bad argument type.
+  let archivePath = archivePathOrBuffer;
+  if (Buffer.isBuffer(archivePathOrBuffer)) {
+    archivePath = archivePathOrBuffer.toString();
+  }
+  if (typeof archivePath !== 'string') return { isAsar: false };
+
+  return asar.splitPath(path.normalize(archivePath));
+};
+
+// Convert asar archive's Stats object to fs's Stats object.
+let nextInode = 0;
+
+const uid = process.getuid != null ? process.getuid() : 0;
+const gid = process.getgid != null ? process.getgid() : 0;
+
+const fakeTime = new Date();
+const msec = (date: Date) => (date || fakeTime).getTime();
+
+const asarStatsToFsStats = function (stats: any) {
+  const { Stats, constants } = require('fs');
+
+  let mode = constants.S_IROTH ^ constants.S_IRGRP ^ constants.S_IRUSR ^ constants.S_IWUSR;
+
+  if (stats.isFile) {
+    mode ^= constants.S_IFREG;
+  } else if (stats.isDirectory) {
+    mode ^= constants.S_IFDIR;
+  } else if (stats.isLink) {
+    mode ^= constants.S_IFLNK;
+  }
+
+  return new Stats(
+    1, // dev
+    mode, // mode
+    1, // nlink
+    uid,
+    gid,
+    0, // rdev
+    undefined, // blksize
+    ++nextInode, // ino
+    stats.size,
+    undefined, // blocks,
+    msec(stats.atime), // atim_msec
+    msec(stats.mtime), // mtim_msec
+    msec(stats.ctime), // ctim_msec
+    msec(stats.birthtime) // birthtim_msec
```

Comment:
`stats.atime` and friends are never actually set, so this always returns fakeTime

Reply:
Trying to avoid changing implementation during ts stuff, also why I had to type a few things as `any`.  Turns our there's a lot of inaccuracies in this file.  They can all be addressed individually once this has landed.